go/chat/attachment/crypto.go

```diff
@@ -0,0 +1,462 @@
+// the encryption key is easy. We don't need to worry about whether the
+// signature might leak the encryption key either, because the signature
+// gets encrypted.
+//
+// Apart from these signing gymnastics, all the large-encrypted-message
+// considerations from https://www.imperialviolet.org/2015/05/16/aeads.html
+// apply here. Namely we use a chunk number to prevent reordering, and we
+// use an empty chunk at the end to detect truncation. A globally unique
+// nonce (for encryption *and* signing) prevents chunk swapping in between
+// messages, and is required for encryption in any case. (It's expected
+// that the chat client will pass in all zeroes for the nonce, because both
+// keys are one-time-use. That's up to the client. G-d help us if we ever
+// reuse those keys.) We also follow the "prefix signatures with an ASCII
+// context string and a null byte" recommendation from
+// https://www.ietf.org/mail-archive/web/tls/current/msg14734.html.
+
+package attachment
+
+import (
+	"crypto/sha512"
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/agl/ed25519"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+type AttachmentNonce *[AttachmentNonceSize]byte
+type SecretboxKey *[SecretboxKeySize]byte
+type SecretboxNonce *[SecretboxNonceSize]byte
+type SignKey *[ed25519.PrivateKeySize]byte
+type VerifyKey *[ed25519.PublicKeySize]byte
+
+const AttachmentNonceSize = 16
+const SecretboxKeySize = 32
+const SecretboxNonceSize = 24
+const SignaturePrefix = "keybase chat attachment\x00"
+const PlaintextChunkLength = 1048576 // 2^20
+const Bin32Tag = 0xc6
+const Bin32Overhead = 5 // The bin32 tag plus 4 length bytes
+const PacketLength = PlaintextChunkLength + ed25519.SignatureSize + secretbox.Overhead + Bin32Overhead
+
+// ===================================
+// single packet encoding and decoding
+// ===================================
+
+func makeChunkNonce(attachmentNonce AttachmentNonce, chunkNum uint64) SecretboxNonce {
+	var ret [SecretboxNonceSize]byte
+	copy(ret[0:16], attachmentNonce[:])
+	var chunkNumBytes [8]byte
+	binary.BigEndian.PutUint64(chunkNumBytes[:], chunkNum)
+	copy(ret[16:24], chunkNumBytes[:])
+	return &ret
+}
+
+func makeSignatureInput(plaintext []byte, encKey SecretboxKey, chunkNonce SecretboxNonce) []byte {
+	chunkHash := sha512.Sum512(plaintext)
+	var ret []byte
+	ret = append(ret, SignaturePrefix...)
+	ret = append(ret, encKey[:]...)
+	ret = append(ret, chunkNonce[:]...)
+	ret = append(ret, chunkHash[:]...)
+	return ret
+}
+
+// When serializing ciphertext chunks, we always use the bin32 MessagePack
+// format, regardless of the size of the chunk. That lets us predict the
+// ciphertext size given the plaintext size.
+func packCiphertext(ciphertext []byte) []byte {
+	if len(ciphertext) > math.MaxUint32 {
+		panic("ciphertext chunk is too long")
+	}
+	len32 := uint32(len(ciphertext))
+	// A bin32 object is byte 0xc6 followed by a 32-bit big-endian length.
+	packet := []byte{0xc6, 0, 0, 0, 0}
```

Comment:
ok let's get rid of the `0xc6` prefix because I don't like requiring encoding to 32-bits-length even if it's a small integer (the opposite of what we have required elsewhere).

does this mean we should get rid of framing altogether?  ok